### PR TITLE
net: buf: linearize: Never return -ENOMEM, just do what user asked to

### PR DESCRIPTION
**Pull request by pfalcon**
Don't try to find "errors" in the values of dst_len and len params
passed to net_buf_linearize(). Instead, do what entails with the
common sense from the values passed in, specifically:

1. Never read more than dst_len (or it would lead to buffer
overflow).
2. It's absolutely ok to read than specified by "len" param, that's
why this function returns number of bytes read in the first place.

The motivation for this change is that it's not useful with its
current behavior. For example, a number of Ethernet drivers linearize
a packet to send, but each does it with its own duplicated adhoc
routine, because net_buf_linearize() would just return error for the
natural use of:

net_buf_linearize(buf, sizeof(buf), pkt->frags, 0, sizeof(buf));

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

### CODE REVIEW

**Comment by pfalcon** [comment]
Currently posted to ask CI's opinion.

**Comment by Vudentz** [comment]
I don't thing using #if 0 is desirable in any circustance, or perhaps you are just using it in order to highlight the code that should be removed? Other than that this looks good to me, in fact even if there were legitimate reason to check for no memory that should probably be converted to assert since that is an obvious bug in the code.

**Comment by pfalcon** [comment]
@Vudentz: Ack, posted just to see if there would be any CI failures, and get any possible quick comments from reviewers. Seeing no failures/immediate objections, will definitely update to properly remove the affected code.

Thanks for looking! Happy New Year!

File: subsys/net/buf.c [hunk 1]
```diff
@@ -681,9 +681,12 @@ int net_buf_linearize(void *dst, size_t dst_len, struct net_buf *src,
 	size_t to_copy;
 	size_t copied;
 
+	len = min(len, dst_len);
+#if 0
```
Comment (tbursztyka) [inline subsys/net/buf.c]: why not just removing the old code instead?

**Review by tbursztyka — DISMISSED** [review]
net_buf.h documentation would need to be updated as well.

**Comment by pfalcon** [comment]
@Vudentz, @tbursztyka: Finished refactoring, please re-review. Suggestions for squashing commits together are accepted.

**Comment by pfalcon** [comment]
Added tests for the updated function behavior.

**Comment by pfalcon** [comment]
@tbursztyka : Ping

File: include/net/net_pkt.h [hunk 2]
```diff
@@ -1122,18 +1122,16 @@ int net_frag_linear_copy(struct net_buf *dst, struct net_buf *src,
...
- * @param offset Starting point to copy from
+ * @param dst_len Destination buffer length
+ * @param src Source packet with fragmented net_buf chain
+ * @param offset Starting offset to copy from
  * @param len Number of bytes to copy
- * @return number of bytes copied if everything is ok
- * @return -ENOMEM on error
+ * @return number of bytes actually copied
  */
 int net_frag_linearize(u8_t *dst, size_t dst_len,
```
Comment (tbursztyka) [inline include/net/net_pkt.h]: Shouldn't it return size_t as well then?

Comment (jhedberg) [reply include/net/net_pkt.h]: ssize_t then, I suppose, since it needs to return negative values as well.

Comment (jhedberg) [reply include/net/net_pkt.h]: Nevermind, I see the negative errors are removed by this PR

**Review by jhedberg — APPROVED** [review]

**Review by tbursztyka — APPROVED** [review]

**Comment by pfalcon** [comment]
Thanks for the reviews. I squashed commits down to 3 after all, so now there's a commit to update net_buf_linearize(), a commit to update net_frag_linearize(), and a commit to add more tests for net_buf_linearize().

File: subsys/net/buf.c [hunk 3]
```diff
@@ -674,16 +674,14 @@ struct net_buf *net_buf_frag_del(struct net_buf *parent, struct net_buf *frag)
...
 	struct net_buf *frag;
 	size_t to_copy;
 	size_t copied;
 
-	if (dst_len < (size_t)len) {
-		return -ENOMEM;
-	}
+	len = min(len, dst_len);
 
 	frag = src;
```
[the remark this answers is not in the document]
Comment (pfalcon) [reply subsys/net/buf.c]: > Unrelated to this patch ...

Ah, don't ask, I see at least one another "unrelated" thing to this patch. Step by step...

File: include/net/buf.h [hunk 4]
```diff
@@ -1300,21 +1300,22 @@ struct net_buf *net_buf_frag_del(struct net_buf *parent, struct net_buf *frag);
 #endif
 
 /**
- * @brief Copy len bytes from src starting from offset to dst buffer
+ * @brief Copy len bytes from net_buf chain starting at offset to linear buffer
```
Comment (nashif) [inline include/net/buf.h]: remove len here -> "Copy bytes ..."

**Review by jukkar — APPROVED** [review]

**Pull request merged**